### ISSUE TITLE
InvalidReturnStatement / InvalidReturnType false positive when array key is a function call

### ISSUE THREAD

**Issue by kylekatarnls** [vimeo/psalm#11857]
In the code below, `foo` cannot return `null`, but Psalm still raises [`NullableReturnStatement`](https://psalm.dev/139) and [`InvalidNullableReturnType`](https://psalm.dev/144)

```php
/** @param array<string, ?string> $foo */
function foo(array $foo, string $key, string $default): string {
    return $foo[bar($key)] ?? $default;
}

function bar(string $key): string {
    return "aa$key";
}
```

https://psalm.dev/r/f8585fb593

Possibly a common root cause with #8213 and/or #10233

**Comment by HenkPoley** [comment]
Had Codex with GPT 5.5 take a look. Looks like a small, but difficult to comprehend fix in `\Psalm\Internal\Analyzer\Statements\Expression\BinaryOp\CoalesceAnalyzer::analyze()`

ExpressionIdentifier::getExtendedVarId() is used to find out whether Psalm can give the entire left-hand array lookup a stable internal id. If it can, normal isset() reconciliation is enough. If it cannot, as with `bar($key)`, then the id `=== null`, and we create a temporary variable so Psalm can narrow that temporary instead

```diff
diff --git a/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/CoalesceAnalyzer.php b/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/CoalesceAnalyzer.php
index e50f0bb55..ab8ee1866 100644
--- a/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/CoalesceAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/CoalesceAnalyzer.php
@@ -6,6 +6,7 @@ namespace Psalm\Internal\Analyzer\Statements\Expression\BinaryOp;
 
 use PhpParser;
 use Psalm\Context;
+use Psalm\Internal\Analyzer\Statements\Expression\ExpressionIdentifier;
 use Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer;
 use Psalm\Internal\Analyzer\StatementsAnalyzer;
 use Psalm\Node\Expr\VirtualIsset;
@@ -45,6 +46,12 @@ final class CoalesceAnalyzer
             || $root_expr instanceof PhpParser\Node\Expr\NullsafePropertyFetch
             || $root_expr instanceof PhpParser\Node\Expr\NullsafeMethodCall
             || $root_expr instanceof PhpParser\Node\Expr\Ternary
+            || ($left_expr instanceof PhpParser\Node\Expr\ArrayDimFetch
+                && ExpressionIdentifier::getExtendedVarId(
+                    $left_expr,
+                    $statements_analyzer->getFQCLN(),
+                    $statements_analyzer,
+                ) === null)
         ) {
             $left_var_id = '$<tmp coalesce var>' . (int) $left_expr->getAttribute('startFilePos');
 
```
